### ISSUE TITLE
Viewing/editing mode-setting should not affect the JSON object

### ISSUE THREAD

**Issue by zehawki** [josdejong/svelte-jsoneditor#72]
This is a json editor, hence the only thing it should ingest and spit out is a JSON object. Whether one wants to view in so called "code" mode, or "tree" mode is only an aspect of viewing / editing preference. Similarly whether one in "formatted" or "compacted" mode. These are just temporary settings and should not change the underlying information.

1. When the `onChange` fires, its very confusing to get either `text` or `json` **depending on which mode the user is in**. Ideally it should be a JSON object always, and it could be based on what the input is, but I dont see any reason to change this based on user's setting.

2. Now here if the `text` key would give back stringified JSON, then also it may be fine, and one can perform a transformation accordingly. But it gives back escaped plain text \n etc `text: "[\n  {\n    \"label\":` 

Whats the decision or purpose here of returning plain text, without the json? If someone wants to escape etc, that should be their work outside the library. Why send the formatting back and then escape everything?

**Comment by zehawki** [comment]
More points;
1. Output from the editor is not meant for display, but for storage. Hence it even more doesnt make sense to get spaces and tabs etc when sending into storage. 
2. Due to this, the text output size in code-format is 2X the size of code-compact.

**Comment by josdejong** [comment]
The reason for defining the contents of the editor as one of `{ json: ..., text: undefined }` or `{ json: undefined, text: '...' }` is that it is possible to have contents that are not valid JSON, and cannot be represented as a JSON. This happens continuously whilst you are typing in code mode: temporarily, your JSON will be invalid when you haven't typed all characters yet like the closing `"` of your string and the comma "," after the array item. Also in tree mode it can happen that the contents are invalid JSON, you can try out the following: open an empty editor in tree mode, so you see the welcome screen. Click inside the editor and then paste the following contents `[1,2,`. Now this is invalid JSON. The editor will fix this for you and inform you about the fix, but it also allows you to manually fix it by going to code mode. This is possible because the internal state can hold invalid JSON as text. Without this, it would be impossible to persist the state of the editor whilst code mode contains invalid JSON, or to open a broken JSON document that you would like to inspect and repair, which would be really bad.

Now the editor is as lazy as possible regarding `onChange` callbacks. Whilst in code mode the editor will always return `{ text: ''...' }`, also if the contents are valid JSON. This is very performant. Alternatively, for every change (very debounced key press), the editor could try to parse the document in to JSON and if successful, return `{json: ...}`. That is really performance heavy though. And besides that, the editor cannot know whether the user has specific formatting that it wants to maintain (compacted, formatted, or something in between), and parsing it will throw away any formatting. 

So to summarize, from a user point of view I want to (a) be able to persist and reload my document at _any_ arbitrary moment, and, (b) when I reload I want the document to be _exactly_ same as how I left it.

> Due to this, the text output size in code-format is 2X the size of code-compact.

What you can do is try to parse into JSON yourself:

```js
onChange(content) {
  if (typeof content.text === 'string') {
    try {
      const parsedContent = {
        json: JSON.parse(content.text) // this can throw an exception
      }
      saveMyDocument(parsedContent )
    } catch (err) {
      // ignore, will save as text anyway
    }
  }
  
  saveMyDocument(content)
}
```

**Comment by josdejong** [comment]
👍  thanks for thinking along
